File: remotedsh-contract/go/fingerprint.go
// a=fingerprint 归一与选取（契约 §6.1，逐字实现；RFC 8122 §5.1 多行）
// 无 sha-256 行 → fail-closed（返回 ""，调用方 MUST 中止，不得继续 SetRemoteDescription）
package contract

import "strings"

// NormalizeFingerprintLine：'a=fingerprint:sha-256 AA:BB:…' → 'aabb…'（64 hex 小写）或 ""
func NormalizeFingerprintLine(line string) string {
	rest, ok := strings.CutPrefix(strings.TrimSpace(line), "a=fingerprint:")
	if !ok {
		// 容忍 'a= fingerprint:' 的空白变体极罕见；按 §6.1 严格前缀处理即可
		return ""
	}
	sp := strings.IndexAny(rest, " \t")
	if sp < 0 {
		return ""
	}
	hashFunc := strings.ToLower(rest[:sp])
	if hashFunc != "sha-256" {
		return ""
	}
	value := rest[sp+1:]
	var b strings.Builder
	for _, r := range value {
		switch {
		case r == ':' || r == ' ' || r == '\t':
			// 删全部冒号与空白
		case r >= 'A' && r <= 'Z':
			b.WriteRune(r + 32) // ASCII 小写
		default:
			b.WriteRune(r)
		}
	}
	return b.String()
}

// ExtractPinnedFingerprint：从 offer SDP 行中取归一指纹。
// 多行时取 sha-256 行；两级冲突以 media 级为准；media 无则取 session 级；都没有 → ""（fail-closed）。
func ExtractPinnedFingerprint(offerLines []string) string {
	var session, media string
	inMedia := false
	for _, line := range offerLines {
		if strings.HasPrefix(line, "m=") {
			inMedia = true
			continue
		}
		fp := NormalizeFingerprintLine(line)
		if fp == "" {
			continue
		}
		if inMedia {
			if media == "" {
				media = fp
			}
		} else if session == "" {
			session = fp
		}
	}
	if media != "" {
		return media
	}
	return session // 无 sha-256 行时为 ""
}

// VerifyPin：与本地持久化 fp 按字节比较（§6.1：不是与内存里的 QR 内容比较）
func VerifyPin(persistedFp string, offerLines []string) bool {
	offered := ExtractPinnedFingerprint(offerLines)
	if offered == "" || persistedFp == "" {
		return false
	}
	norm := strings.ToLower(strings.Map(func(r rune) rune {
		if r == ':' || r == ' ' || r == '\t' {
			return -1
		}
		return r
	}, persistedFp))
	return offered == norm
}

File: remotedsh-contract/js/fingerprint.mjs
// a=fingerprint 归一与选取（契约 §6.1 红线，逐字实现；RFC 8122 §5.1 多行）
// 规则：取 hash-func 归一后等于 sha-256 的行（两级冲突 media 级为准）→ 剥前缀 → 删全部冒号与空白 → ASCII 小写 → 按字节比较
// 无 sha-256 行 → fail-closed（返回 null，调用方 MUST 中止，不得继续 setRemoteDescription）

/** 归一单行值：'sha-256 AA:BB:…' → 'aabb…'（64 hex 小写字符）或 null */
export function normalizeFingerprintLine(line) {
  const m = /^\s*a\s*=\s*fingerprint\s*:\s*(\S+)\s+(.+)$/i.exec(line)
  if (!m) return null
  const hashFunc = m[1].toLowerCase() // 'SHA-256' → 'sha-256'
  if (hashFunc !== 'sha-256') return null
  return m[2].replace(/[:\s]+/g, '').toLowerCase() // 删全部冒号与空白 → ASCII 小写
}

/**
 * 从 offer SDP 中取归一指纹。返回 64 字符小写 hex，或 null（fail-closed）。
 * offerLines: SDP 行数组（已按 \r?\n 切分）
 */
export function extractPinnedFingerprint(offerLines) {
  let sessionLevel = null
  let mediaLevel = null
  let inMedia = false
  for (const line of offerLines) {
    if (/^m=/.test(line)) { inMedia = true; continue }
    const fp = normalizeFingerprintLine(line)
    if (fp === null) continue
    if (inMedia) mediaLevel = mediaLevel ?? fp // 冲突时以 media 级为准（§6.1）
    else sessionLevel = sessionLevel ?? fp
  }
  const picked = mediaLevel ?? sessionLevel // 优先 media 级；media 无指纹时取 session 级的 sha-256 行
  return picked ?? null // 无 sha-256 行 → fail-closed
}

/** pin 校验（契约 §6.1）：persistedFp 与 offer 归一结果按字节比较；任一缺失即不匹配 */
export function verifyPin(persistedFp, offerLines) {
  const offered = extractPinnedFingerprint(offerLines)
  if (!offered || !persistedFp) return false
  return offered === String(persistedFp).replace(/[:\s]+/g, '').toLowerCase()
}
